### ISSUE TITLE
BCEWithLogitsLoss: Check if labels within zero and one

### ISSUE THREAD

**Issue by sehoffmann** [pytorch/pytorch#110865]
### 🚀 The feature, motivation and pitch

`BCEWithLogitsLoss` currently performs no kind of checking wrt. the targets whatsoever. This means, that calling the loss function with targets not contained within the interval [0,1] succesfully computes values without throwing an error. However, these values don't have any kind of meaning.

In practice, though, passing invalid targets to `BCEWithLogitsLoss` can easily happen and can go unnoticed! Such mistakes can render whole experiments invalid; **if** they are even noticed in the first place!

For instance, this can happen if multi-class labels and predictions (i.e. 0,1,2,...,n) are erroneously used with BCEWithLogitsLoss. *Shouldn't this result in a shape-mismatch due to the prediction being a categorical distribution?* No, both predictions and targets will be broadcastable to each other in that case. In fact, commonly, their shape will be identical.

**Comment by mikaylagawarecki** [comment]
Since we validate target for `BCELoss` it seems reasonable to add the same validation for `BCEWithLogitsLoss`. I could review a PR that adds the checks

Here is how the checks are added on CPU
https://github.com/pytorch/pytorch/blob/ed2cc4dd599316f1270459897ab1e5a6d178a617/aten/src/ATen/native/Loss.cpp#L285-L288

and on CUDA
https://github.com/pytorch/pytorch/blob/ed2cc4dd599316f1270459897ab1e5a6d178a617/aten/src/ATen/native/cuda/Loss.cu#L95
